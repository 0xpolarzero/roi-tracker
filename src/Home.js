import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';

const Home = () => {
  //

  return (
    <div className='home'>
      <Link to='/'>
        <h1>Go to app</h1>
      </Link>
    </div>
  );
};

export default Home;
