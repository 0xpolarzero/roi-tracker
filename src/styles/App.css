:root {
  --purple-main: rgb(160, 108, 213);
  --purple-light: rgb(184, 147, 221);
  --purple-dark: rgb(130, 71, 190);
  --black-main: #141414;
  --black-light: #252526;
  --main-font: 'Plus Jakarta Sans', -apple-system, BlinkMacSystemFont,
    'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans',
    'Droid Sans', 'Helvetica Neue', sans-serif;
}

body {
  margin: 0;
  font-family: var(--main-font);
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

input {
  background: rgba(255, 255, 255, 0.1);
  color: #fff;
  font-family: var(--main-font);
  border: 0;
  border-radius: 3px;
  padding: 0.3rem 0.5rem;
}

input:active {
  outline: none;
}

input::placeholder {
  color: rgba(255, 255, 255, 0.5);
}

button {
  background: rgba(255, 255, 255, 0.3);
  color: #fff;
  font-family: var(--main-font);
  font-size: 1rem;
  font-weight: 400;
  padding: 0.3rem 2rem;
  border-radius: 6px;
  box-shadow: rgba(0, 0, 0, 0.12) 0px 1px 3px, rgba(0, 0, 0, 0.24) 0px 1px 2px;
  outline: none;
  border: 0;
  cursor: pointer;

  transition: all 0.2s linear;
}

button:hover {
  background: rgba(255, 255, 255, 0.1);
  box-shadow: rgba(0, 0, 0, 0.06) 0px 2px 4px 0px inset;
}

ul {
  list-style: none;
  padding: 0;
  margin: 0;
}

.App {
  position: relative;

  color: white;
  min-height: 100vh;
  display: grid;
  padding: 1rem 2rem;
}

.bg-blur {
  position: absolute;
  width: 100%;
  height: 100%;
  /* background: rgb(20, 20, 20);
  background: linear-gradient(
    180deg,
    #141414 0%,
    #8247be 42%,
    #ff007b 100%
  ); */
  background: hsla(270, 48%, 51%, 1);

  background: radial-gradient(
    circle,
    hsla(270, 48%, 51%, 1) 0%,
    hsla(0, 0%, 8%, 1) 100%
  );
  /* background-size: 120% 120%; */

  z-index: -2;
}

.notif {
  background: rgba(255, 255, 255, 0);
  position: absolute;
  color: #ffffff;
  bottom: 1rem;
  right: 1rem;
  padding: 0.6rem 2rem;
  border-radius: 6px;
  opacity: 0;
  pointer-events: none;

  transition: opacity 0.5s, background 0.5s;
}

.notif.error,
.notif.info,
.notif.hint {
  opacity: 1;
}

.notif.error {
  background: rgba(166, 66, 83, 0.8);
}

.notif.info {
  background: rgba(97, 201, 97, 0.8);
}

.notif.hint {
  background: rgba(255, 186, 8, 0.8);
}

#popup {
  position: absolute;
  top: 0;
  left: 2rem;
  font-family: var(--main-font);
  font-size: 0.8rem;
  font-weight: 300;
  color: #fff;
  background: rgba(255, 255, 255, 0);
  border: 1px solid rgba(255, 186, 8, 0.6);
  border-radius: 1rem;
  width: 200px;
  padding: 0.5rem;
}

.separator-hor {
  border-bottom: 1px solid rgba(255, 255, 255, 0.1);
}

.separator-ver {
  border-right: 1px solid red;
}
