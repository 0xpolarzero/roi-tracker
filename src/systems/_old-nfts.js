import { createAlchemyWeb3 } from '@alch/alchemy-web3';

const provider = createAlchemyWeb3(`https://eth-mainnet.alchemyapi.io/v2/demo`);

async function getNftExchanges(startBlock, endBlock, addresses) {
  const transactions = await getNftTransactions(
    startBlock,
    endBlock,
    addresses,
  );
  // Remove a transaction if it went in & out during the period
    const exchanges = transactions.in.filter((tx) => {
        return transactions.out.find((outTx) => {
            return tx.id === outTx.id;
        });
    }))

  console.log(exchanges);
}

async function getNftTransactions(startBlock, endBlock, addresses) {
  let exchanges = { in: [], out: [] };

  for (const address of addresses) {
    const exchangesInForAddress = await provider.alchemy.getAssetTransfers({
      fromBlock: startBlock.block,
      toBlock: endBlock.block,
      toAddress: address,
      order: 'desc',
      category: ['erc721', 'erc1155'],
      withMetaData: true,
    });

    const exchangesOutForAddress = await provider.alchemy.getAssetTransfers({
      fromBlock: startBlock.block,
      toBlock: endBlock.block,
      fromAddress: address,
      order: 'desc',
      category: ['erc721', 'erc1155'],
      withMetaData: true,
    });

    exchanges.in.push(...exchangesInForAddress.transfers);
    exchanges.out.push(...exchangesOutForAddress.transfers);

    return exchanges;
  }
}

export { getNftExchanges };
