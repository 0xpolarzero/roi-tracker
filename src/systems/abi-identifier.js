import { Interface } from '@ethersproject/abi';

async function identifyAction(transaction) {
  console.log(Interface);
}

export { identifyAction };
