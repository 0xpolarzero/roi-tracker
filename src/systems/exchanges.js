import { fetchData } from '../utils/utils';

const getTransactions = async (provider, address, block) => {
  // Get a list of transactions made by this address during this period
  const transactions = await provider.alchemy.getAssetTransfers({
    fromAddress: address,
    fromBlock: block,
  });

  // Get the addresses of all major cryptocurrency exchanges
  const exchanges = await getExchangeAddresses();
};

const getExchangeAddresses = async () => {
  const url =
    'https://etherscan.io/accounts/label/exchange?subcatid=undefined&size=400&start=0&col=1&order=asc';
  const url2 = 'https://en.wikipedia.org/wiki/Cricket_World_Cup';

  const response = await getRawData(url2);
  console.log(response);
  console.log('test');
};

const getRawData = (URL) => {
  return fetch(URL)
    .then((response) => response.text())
    .then((data) => {
      return data;
    });
};

export { getExchangeAddresses };
