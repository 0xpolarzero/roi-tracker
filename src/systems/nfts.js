import { createAlchemyWeb3 } from '@alch/alchemy-web3';

import { displayNotif } from './utils';

const web3 = createAlchemyWeb3(`https://eth-mainnet.alchemyapi.io/v2/demo`);

async function getNftExchanges(provider, startBlock, endBlock, addresses) {
  const currentNfts = await getCurrentNfts(provider, addresses).catch((err) => {
    console.log(err);
    displayNotif('error', 'We could not load your current NFTs.', 4000);
  });

  const transactions = await getNftTransactions(
    provider,
    startBlock,
    endBlock,
    addresses,
  ).catch((err) => {
    console.log(err);
    displayNotif(
      'error',
      'There was an issue fetching the NFT exchanges with your addresses.',
      4000,
    );
  });

  // Exclude local transactions corresponding to a transfer between addresses
  if (transactions) {
    transactions.filter((txn) => {
      return !(addresses?.includes(txn.from) && addresses?.includes(txn.to));
    });
  }
}

async function getCurrentNfts(provider, addresses) {
  let currentNfts = [];

  for (const address of addresses) {
    const currentNft = await provider.account.getNFTs({
      address: address,
    });
    currentNfts.push(...currentNft.result);
  }

  return currentNfts;
}

async function getNftTransactions(provider, startBlock, endBlock, addresses) {
  let exchanges = [];

  for (const address of addresses) {
    const exchangesInForAddress = await web3.alchemy.getAssetTransfers({
      fromBlock: startBlock.block,
      toBlock: endBlock.block,
      toAddress: address,
      order: 'desc',
      category: ['erc721', 'erc1155'],
      withMetaData: true,
    });

    const exchangesOutForAddress = await web3.alchemy.getAssetTransfers({
      fromBlock: startBlock.block,
      toBlock: endBlock.block,
      fromAddress: address,
      order: 'desc',
      category: ['erc721', 'erc1155'],
      withMetaData: true,
    });

    exchanges.push(
      ...exchangesInForAddress.transfers,
      ...exchangesOutForAddress.transfers,
    );
  }

  return exchanges;
}

export { getNftExchanges };
