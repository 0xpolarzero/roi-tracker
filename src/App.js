import React, { useEffect, useState } from 'react';

import { createAlchemyWeb3 } from '@alch/alchemy-web3';
import { useMoralis } from 'react-moralis';
import EthDater from 'ethereum-block-by-date';

import Tracker from './components/Tracker';
import Connect from './components/Connect';

import { displayNotif, fetchData } from './systems/utils';

import './styles/index.css';

const web3 = createAlchemyWeb3(
  `https://eth-mainnet.alchemyapi.io/v2/${process.env.REACT_APP_ALCHEMY_API_KEY}`,
);

const dater = new EthDater(web3);

const App = () => {
  const [ethPriceValue, setEthPriceValue] = useState(0);
  const [isLogged, setIsLogged] = useState(false);

  const { isWeb3Enabled } = useMoralis();

  // Display ETH Price
  useEffect(() => {
    let interval;
    const setEthPrice = async () => {
      interval = setInterval(async () => {
        const ethPrice = await getEthPrice();
        setEthPriceValue(ethPrice);
      }, 10000);
      const ethPrice = await getEthPrice();
      setEthPriceValue(ethPrice);
    };
    setEthPrice();

    return () => clearInterval(interval);
  }, []);

  const getEthPrice = async () => {
    const ethPrice = await fetchData(
      'https://api.coinbase.com/v2/exchange-rates?currency=ETH',
    ).catch((err) => {
      console.log(err);
      displayNotif('error', 'Failed to fetch Ether price.', 2000);
      return 'Unknown';
    });

    return Number(ethPrice.data.rates.USD).toFixed(2);
  };

  useEffect(() => {
    window.ethereum.on('accountsChanged', updateWalletStatus);
    updateWalletStatus();
  }, []);

  const updateWalletStatus = async () => {
    const accounts = await web3.eth.getAccounts();
    setIsLogged(accounts.length > 0);
  };

  // Gérer la connextion dans Header
  // Dès qu'elle est faite lancer connectWallet() avec l'address et les tokens dans le state + isLogged
  // Envoyer ça dans Tracker
  // Ajouter ces éléments dans la prise en compte (par exemple if (isLogged) { addAddress(addressFromWallet) } })

  return (
    <div className='App'>
      <Tracker
        web3={web3}
        dater={dater}
        ethPriceValue={ethPriceValue}
        isLogged={isLogged}
      />
      <div className='notif'></div>
      <div className='bg-blur'></div>
    </div>
  );
};

export default App;
