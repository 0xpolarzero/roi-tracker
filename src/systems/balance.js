import { getExchangeAddresses } from './exchanges/exchanges';
import { TimestampConverter } from './timestamp';

// const tokenSource = 'https://tokens.coingecko.com/uniswap/all.json';
// https://solveforum.com/forums/threads/solved-get-token-balance-of-address-at-particular-block-number-with-alchemy.609041/

async function getEthBalance(provider, block, addresses) {
  let balancesCombined = 0;

  for (const address of addresses) {
    const balanceFromAddress = await provider.eth
      .getBalance(address.toString(), block)
      .catch((err) => {
        console.log(err);
      });
    const formatted = provider.utils.fromWei(balanceFromAddress, 'ether');
    balancesCombined += Number(formatted);
  }

  return balancesCombined;
}

async function getTokenBalance(provider, block, walletAddresses, tokenAddress) {
  let balancesCombined = 0;

  const contract = new provider.eth.Contract(minABI, tokenAddress);

  for (const address of walletAddresses) {
    const tokenBalanceFromAddress = await contract.methods
      .balanceOf(address)
      .call(null, block);
    // Format balance in ether
    const formatted = provider.utils.fromWei(tokenBalanceFromAddress, 'ether');
    balancesCombined += Number(formatted);
  }

  return balancesCombined;
}

const minABI = [
  // balanceOf
  {
    constant: true,
    inputs: [{ name: '_owner', type: 'address' }],
    name: 'balanceOf',
    outputs: [{ name: 'balance', type: 'uint256' }],
    type: 'function',
  },
];

const isValidAddress = (provider, address) => {
  const isValid = provider.utils.isAddress(address);
  return isValid;
};

export { getEthBalance, getTokenBalance, isValidAddress };
