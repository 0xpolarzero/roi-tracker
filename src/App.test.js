import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { BrowserRouter } from 'react-router-dom';
import { WagmiConfig } from 'wagmi';
import { setupClient } from './systems/wagmi-client-setup';

import App from './App';
import Home from './Home';

const Wrapper = ({ children }) => {
  return (
    <WagmiConfig client={setupClient()}>
      <BrowserRouter>{children}</BrowserRouter>
    </WagmiConfig>
  );
};

const renderWithRouting = (ui) => {
  return render(ui, { wrapper: Wrapper });
};

describe('Home component', () => {
  it('renders home', () => {
    // since screen does not have the container property, we'll destructure render to obtain container for this test
    const { container } = renderWithRouting(<Home />);
    expect(container).toMatchSnapshot();
  });
  it('renders correct heading', () => {
    const { getByRole } = renderWithRouting(<Home />);
    expect(getByRole('heading').textContent).toMatch(/go to app/i);
  });
});

describe('App component', () => {
  it('renders the app', () => {
    // since screen does not have the container property, we'll destructure render to obtain container for this test
    const { container } = renderWithRouting(<App />);
    expect(container).toMatchSnapshot();
  });
});

describe('Wallet connection buttons', () => {
  it('renders Metamask connector', () => {
    renderWithRouting(<App />);
    // Get all the buttons
    const buttons = screen.getAllByRole('button');
    // Check if there is a button with the text 'Connect'
    expect(buttons[0].textContent).toMatch(/Metamask/i);
  });
  it('renders Coinbase connector', () => {
    renderWithRouting(<App />);
    // Get all the buttons
    const buttons = screen.getAllByRole('button');
    // Check if there is a button with the text 'Connect'
    expect(buttons[1].textContent).toMatch(/Coinbase Wallet/i);
  });
  it('renders WalletConnect connector', () => {
    renderWithRouting(<App />);
    // Get all the buttons
    const buttons = screen.getAllByRole('button');
    // Check if there is a button with the text 'Connect'
    expect(buttons[2].textContent).toMatch(/WalletConnect/i);
  });
});

describe('Wallet connection system', () => {
  it('connects with Metamask', () => {
    renderWithRouting(<App />);
    // Get all the buttons
    const metamaskButton = screen.getByRole('button', { name: /Metamask/i });
    // Click the button with the text 'Connect'
    userEvent.click(metamaskButton);
    // Check if the wallet is connected
    expect(metamaskButton.textContent).toMatch('MetaMask (unsupported)');
  });
});
