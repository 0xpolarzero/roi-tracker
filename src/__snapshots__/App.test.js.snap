// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders the app 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="connect-wallet"
    >
      <div
        class="header"
      >
        Connect your wallet to launch the app
      </div>
      <div
        class="connectors"
      >
        <button
          disabled=""
        >
          MetaMask
           (unsupported)
        </button>
        <button>
          Coinbase Wallet
        </button>
        <button>
          WalletConnect
        </button>
        <button
          disabled=""
        >
          Injected
           (unsupported)
        </button>
      </div>
    </div>
    <div
      class="notif"
    />
    <div
      class="bg-blur"
    />
  </div>
</div>
`;

exports[`Home component renders home 1`] = `
<div>
  <div
    class="home"
  >
    <a
      href="/"
    >
      <h1>
        Go to app
      </h1>
    </a>
  </div>
</div>
`;
